ante/eth.go

```diff
@@ -290,20 +291,37 @@ func CheckAndSetEthSenderNonce(
 				"account %s is nil", common.BytesToAddress(from.Bytes()),
 			)
 		}
-		nonce := acc.GetSequence()
+		expectedNonce := acc.GetSequence()
+		txNonce := tx.Nonce()
+		fromStr := from.String()
 
+		// if flag is set, we bypass nonce all check verification
 		if !unsafeUnOrderedTx {
-			// we merged the nonce verification to nonce increment, so when tx includes multiple messages
-			// with same sender, they'll be accepted.
-			if tx.Nonce() != nonce {
+			// to support tx replacement, we check if the transaction nonce exists in the cache and if yes we skip
+			// nonce verification, and we don't set the sequence
+			// We allow skip verification only during CheckTx to keep sequence safe during the execution.
+			if ctx.IsCheckTx() && !ctx.IsReCheckTx() && cache.Exists(fromStr, txNonce) {
+				continue
+			}
+
+			// nonce verification, the sequence needs to be in order
+			if txNonce != expectedNonce {
 				return errorsmod.Wrapf(
 					errortypes.ErrInvalidSequence,
-					"invalid nonce; got %d, expected %d", tx.Nonce(), nonce,
+					"invalid nonce; got %d, expected %d", txNonce, expectedNonce,
 				)
 			}
+
```

Comment:
```suggestion
if !unsafeUnOrderedTx {
			ex := cache.Exists(fromStr, txNonce)
			// to support tx replacement, we check if the transaction nonce exists in the cache and if yes we skip
			// nonce verification, and we don't set the sequence
			// We allow skip verification only during CheckTx to keep sequence safe during the execution.
			if ctx.IsCheckTx() && !ctx.IsReCheckTx() && ex {
				continue
			}

			// nonce verification, the sequence needs to be in order
			if txNonce != expectedNonce {
				return errorsmod.Wrapf(
					errortypes.ErrInvalidSequence,
					"invalid nonce; got %d, expected %d", txNonce, expectedNonce,
				)
			}
			isDeliverTx := !ctx.IsCheckTx()

			if isDeliverTx && ex {
				cache.Delete(fromStr, txNonce)
			} else {
				cache.Set(fromStr, txNonce)
			}

		}
```

1. check cache exists once
2. make it clear that only if cache exists an it is deliver tx should we then delete the cache

Just a suggestion! no issues with the current code